File: backend/internal/anki/collection.go
package anki

import (
	"archive/zip"
	"bytes"
	"database/sql"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"sort"
	"strings"

	"github.com/klauspost/compress/zstd"
)

// readCollection locates the SQLite collection inside the archive and returns its raw
// bytes. Anki has shipped four collection file names over time; the newer ".anki21b"/
// ".anki2b" variants are zstd-compressed, the older ones are plain SQLite. The returned
// string is the matched file name (used as a label in the import report).
func readCollection(zipReader *zip.Reader) ([]byte, string, error) {
	for _, candidate := range []struct {
		name       string
		compressed bool
	}{
		{"collection.anki21b", true},
		{"collection.anki2b", true},
		{"collection.anki21", false},
		{"collection.anki2", false},
	} {
		raw, err := readZipFile(zipReader, []string{candidate.name})
		if err != nil {
			continue
		}
		if !candidate.compressed {
			return raw, candidate.name, nil
		}
		decoder, err := zstd.NewReader(bytes.NewReader(raw))
		if err != nil {
			return nil, candidate.name, err
		}
		defer decoder.Close()
		decoded, err := io.ReadAll(decoder)
		if err != nil {
			return nil, candidate.name, err
		}
		return decoded, candidate.name, nil
	}
	return nil, "", os.ErrNotExist
}

// readMetadata pulls the deck, deck-config and note-type definitions out of the single-row
// "col" table. These are stored as JSON blobs, parsed by the read* helpers below. Decks and
// note types are sorted by name for stable output.
func readMetadata(db *sql.DB, warnings []string) ([]Deck, []DeckConfig, []NoteType, []string) {
	var decksRaw, deckConfigsRaw, modelsRaw string
	row := db.QueryRow(`SELECT decks, dconf, models FROM col LIMIT 1`)
	if err := row.Scan(&decksRaw, &deckConfigsRaw, &modelsRaw); err != nil {
		warnings = append(warnings, "collection metadata unavailable: "+err.Error())
		return nil, nil, nil, warnings
	}

	decks := readDecks(decksRaw)
	deckConfigs := readDeckConfigs(deckConfigsRaw)
	noteTypes := readNoteTypes(modelsRaw)
	sort.Slice(decks, func(i, j int) bool { return decks[i].Name < decks[j].Name })
	sort.Slice(noteTypes, func(i, j int) bool { return noteTypes[i].Name < noteTypes[j].Name })
	return decks, deckConfigs, noteTypes, warnings
}

func readDecks(raw string) []Deck {
	var payload map[string]map[string]any
	if json.Unmarshal([]byte(raw), &payload) != nil {
		return nil
	}
	out := make([]Deck, 0, len(payload))
	for id, deckRaw := range payload {
		name := stringValue(deckRaw["name"])
		// Anki encodes nesting in the name ("Parent::Child"); resolve the parent's ID by
		// matching the prefix name against the other decks.
		parentID := ""
		if strings.Contains(name, "::") {
			parentName := name[:strings.LastIndex(name, "::")]
			for otherID, other := range payload {
				if stringValue(other["name"]) == parentName {
					parentID = otherID
					break
				}
			}
		}
		out = append(out, Deck{
			ID:          id,
			Name:        fallback(name, "Imported Deck"),
			ParentID:    parentID,
			ConfigID:    numberString(deckRaw["conf"]),
			Description: stringValue(deckRaw["desc"]),
			Dynamic:     intValue(deckRaw["dyn"]) != 0,
			Mod:         intValue(deckRaw["mod"]),
			USN:         intValue(deckRaw["usn"]),
			Raw:         deckRaw,
		})
	}
	return out
}

func readDeckConfigs(raw string) []DeckConfig {
	var payload map[string]map[string]any
	if json.Unmarshal([]byte(raw), &payload) != nil {
		return nil
	}
	out := make([]DeckConfig, 0, len(payload))
	for id, cfg := range payload {
		out = append(out, DeckConfig{ID: id, Name: fallback(stringValue(cfg["name"]), "Default"), Raw: cfg})
	}
	sort.Slice(out, func(i, j int) bool { return out[i].ID < out[j].ID })
	return out
}

func readNoteTypes(raw string) []NoteType {
	var payload map[string]map[string]any
	if json.Unmarshal([]byte(raw), &payload) != nil {
		return nil
	}
	out := make([]NoteType, 0, len(payload))
	for id, model := range payload {
		fields := []Field{}
		if rawFields, ok := model["flds"].([]any); ok {
			for i, item := range rawFields {
				fieldRaw, _ := item.(map[string]any)
				fields = append(fields, Field{
					Name:        fallback(stringValue(fieldRaw["name"]), fmt.Sprintf("Field %d", i+1)),
					Ord:         coalesceInt(fieldRaw["ord"], int64(i)),
					Sticky:      boolValue(fieldRaw["sticky"]),
					RTL:         boolValue(fieldRaw["rtl"]),
					Font:        stringValue(fieldRaw["font"]),
					Size:        intValue(fieldRaw["size"]),
					Description: stringValue(fieldRaw["description"]),
				})
			}
		}
		templates := []Template{}
		if rawTemplates, ok := model["tmpls"].([]any); ok {
			for i, item := range rawTemplates {
				templateRaw, _ := item.(map[string]any)
				templates = append(templates, Template{
					Name:   fallback(stringValue(templateRaw["name"]), fmt.Sprintf("Card %d", i+1)),
					Ord:    coalesceInt(templateRaw["ord"], int64(i)),
					QFmt:   stringValue(templateRaw["qfmt"]),
					AFmt:   stringValue(templateRaw["afmt"]),
					DeckID: numberString(templateRaw["did"]),
				})
			}
		}
		out = append(out, NoteType{
			ID:        id,
			Name:      fallback(stringValue(model["name"]), "Imported Note"),
			Type:      intValue(model["type"]),
			CSS:       stringValue(model["css"]),
			LatexPre:  stringValue(model["latexPre"]),
			LatexPost: stringValue(model["latexPost"]),
			Fields:    fields,
			Templates: templates,
			Raw:       model,
		})
	}
	return out
}

File: src/utils/apkg/collection.ts
// Deck / deck-config / note-type metadata readers, mirroring backend/internal/anki/collection.go.
// The "col" table stores these as JSON blobs keyed by id; the readers coerce them into the
// typed model with the exact same fallbacks and orderings as the Go parser.

import type { Database } from "sql.js";
import type { ApkgDeck, ApkgDeckConfig, ApkgField, ApkgNoteType, ApkgTemplate } from "./types";
import {
  boolValue,
  coalesceInt,
  fallback,
  intValue,
  numberString,
  omitEmpty,
  omitFalse,
  omitZero,
  stringValue,
} from "./coerce";

/**
 * byteLess compares strings the way Go compares them (byte order of the UTF-8 encoding).
 * JS's `<` compares UTF-16 code units, which agrees with UTF-8 byte order everywhere except
 * across the surrogate range — irrelevant for real deck names, but the sort must be
 * deterministic and identical to the Go parser's for the parity goldens to hold.
 */
function byteLess(a: string, b: string): boolean {
  return a < b;
}

export interface CollectionMetadata {
  decks: ApkgDeck[] | null;
  deckConfigs: ApkgDeckConfig[] | null;
  noteTypes: ApkgNoteType[] | null;
  warnings: string[];
}

/** readMetadata pulls decks/dconf/models out of the single-row "col" table. */
export function readMetadata(db: Database, warnings: string[]): CollectionMetadata {
  let decksRaw = "";
  let deckConfigsRaw = "";
  let modelsRaw = "";
  try {
    const stmt = db.prepare("SELECT decks, dconf, models FROM col LIMIT 1");
    try {
      if (!stmt.step()) throw new Error("col table is empty");
      const [d, c, m] = stmt.get();
      decksRaw = String(d ?? "");
      deckConfigsRaw = String(c ?? "");
      modelsRaw = String(m ?? "");
    } finally {
      stmt.free();
    }
  } catch (err) {
    warnings.push(`collection metadata unavailable: ${err instanceof Error ? err.message : String(err)}`);
    return { decks: null, deckConfigs: null, noteTypes: null, warnings };
  }

  const decks = readDecks(decksRaw);
  const deckConfigs = readDeckConfigs(deckConfigsRaw);
  const noteTypes = readNoteTypes(modelsRaw);
  decks?.sort((a, b) => (byteLess(a.name, b.name) ? -1 : byteLess(b.name, a.name) ? 1 : 0));
  noteTypes?.sort((a, b) => (byteLess(a.name, b.name) ? -1 : byteLess(b.name, a.name) ? 1 : 0));
  return { decks, deckConfigs, noteTypes, warnings };
}

function parseJSONMap(raw: string): Record<string, Record<string, unknown>> | null {
  try {
    const parsed = JSON.parse(raw);
    if (parsed === null || typeof parsed !== "object" || Array.isArray(parsed)) return null;
    return parsed as Record<string, Record<string, unknown>>;
  } catch {
    return null;
  }
}

export function readDecks(raw: string): ApkgDeck[] | null {
  const payload = parseJSONMap(raw);
  if (!payload) return null;
  const out: ApkgDeck[] = [];
  for (const [id, deckRaw] of Object.entries(payload)) {
    const name = stringValue(deckRaw["name"]);
    // Anki encodes nesting in the name ("Parent::Child"); resolve the parent's id by
    // matching the prefix name against the other decks.
    let parentId = "";
    if (name.includes("::")) {
      const parentName = name.slice(0, name.lastIndexOf("::"));
      for (const [otherId, other] of Object.entries(payload)) {
        if (stringValue(other["name"]) === parentName) {
          parentId = otherId;
          break;
        }
      }
    }
    out.push({
      id,
      name: fallback(name, "Imported Deck"),
      parentId: omitEmpty(parentId),
      configId: omitEmpty(numberString(deckRaw["conf"])),
      description: omitEmpty(stringValue(deckRaw["desc"])),
      dynamic: intValue(deckRaw["dyn"]) !== 0,
      mod: omitZero(intValue(deckRaw["mod"])),
      usn: omitZero(intValue(deckRaw["usn"])),
      raw: deckRaw,
    });
  }
  return out;
}

export function readDeckConfigs(raw: string): ApkgDeckConfig[] | null {
  const payload = parseJSONMap(raw);
  if (!payload) return null;
  const out: ApkgDeckConfig[] = Object.entries(payload).map(([id, cfg]) => ({
    id,
    name: fallback(stringValue(cfg["name"]), "Default"),
    raw: cfg,
  }));
  out.sort((a, b) => (byteLess(a.id, b.id) ? -1 : byteLess(b.id, a.id) ? 1 : 0));
  return out;
}

export function readNoteTypes(raw: string): ApkgNoteType[] | null {
  const payload = parseJSONMap(raw);
  if (!payload) return null;
  const out: ApkgNoteType[] = [];
  for (const [id, model] of Object.entries(payload)) {
    const fields: ApkgField[] = [];
    if (Array.isArray(model["flds"])) {
      (model["flds"] as unknown[]).forEach((item, i) => {
        const fieldRaw = (item ?? {}) as Record<string, unknown>;
        fields.push({
          name: fallback(stringValue(fieldRaw["name"]), `Field ${i + 1}`),
          ord: coalesceInt(fieldRaw["ord"], i),
          sticky: omitFalse(boolValue(fieldRaw["sticky"])),
          rtl: omitFalse(boolValue(fieldRaw["rtl"])),
          font: omitEmpty(stringValue(fieldRaw["font"])),
          size: omitZero(intValue(fieldRaw["size"])),
          description: omitEmpty(stringValue(fieldRaw["description"])),
        });
      });
    }
    const templates: ApkgTemplate[] = [];
    if (Array.isArray(model["tmpls"])) {
      (model["tmpls"] as unknown[]).forEach((item, i) => {
        const templateRaw = (item ?? {}) as Record<string, unknown>;
        templates.push({
          name: fallback(stringValue(templateRaw["name"]), `Card ${i + 1}`),
          ord: coalesceInt(templateRaw["ord"], i),
          qfmt: stringValue(templateRaw["qfmt"]),
          afmt: stringValue(templateRaw["afmt"]),
          deckId: omitEmpty(numberString(templateRaw["did"])),
        });
      });
    }
    out.push({
      id,
      name: fallback(stringValue(model["name"]), "Imported Note"),
      type: intValue(model["type"]),
      css: stringValue(model["css"]),
      latexPre: omitEmpty(stringValue(model["latexPre"])),
      latexPost: omitEmpty(stringValue(model["latexPost"])),
      fields,
      templates,
      raw: model,
    });
  }
  return out;
}
